# 01 - Data Understanding

## Objective

The objective of this notebook is to gain an initial understanding of the dataset by exploring its structure, identifying potential data quality issues, and documenting observations before performing any cleaning or transformation.

### Data Understanding Checklist

- [x] Import libraries
- [x] Load dataset
- [x] Preview data
- [x] Check dataset dimensions
- [x] Inspect data types
- [x] Generate summary statistics
- [x] Analyze missing values
- [x] Check duplicate records
- [x] Explore categorical features
- [x] Document initial findings

## 1. Import Libraries

In [16]:
import pandas as pd
import numpy as np

## 2. Load Data

In [17]:
df = pd.read_excel('../data/online_retail_II.xlsx')

## 3. Dataset Preview

In [18]:
df.head()

,Invoice,StockCode,Description,Quantity,InvoiceDate,Price,Customer ID,Country
0,489434,85048,15CM CHRISTMAS GLASS BALL 20 LIGHTS,12,2009-12-01 07:45:00,6.95,13085.0,United Kingdom
1,489434,79323P,PINK CHERRY LIGHTS,12,2009-12-01 07:45:00,6.75,13085.0,United Kingdom
2,489434,79323W,WHITE CHERRY LIGHTS,12,2009-12-01 07:45:00,6.75,13085.0,United Kingdom
3,489434,22041,"RECORD FRAME 7"" SINGLE SIZE",48,2009-12-01 07:45:00,2.10,13085.0,United Kingdom
4,489434,21232,STRAWBERRY CERAMIC TRINKET BOX,24,2009-12-01 07:45:00,1.25,13085.0,United Kingdom


## 4. Dataset Dimensions

In [19]:
df.shape

(525461, 8)

## 5. Dataset Information

Inspect data types, missing values and memory usage.

In [20]:
df.info()

<class 'pandas.DataFrame'>
RangeIndex: 525461 entries, 0 to 525460
Data columns (total 8 columns):
 #   Column       Non-Null Count   Dtype         
---  ------       --------------   -----         
 0   Invoice      525461 non-null  object        
 1   StockCode    525461 non-null  object        
 2   Description  522533 non-null  object        
 3   Quantity     525461 non-null  int64         
 4   InvoiceDate  525461 non-null  datetime64[us]
 5   Price        525461 non-null  float64       
 6   Customer ID  417534 non-null  float64       
 7   Country      525461 non-null  str           
dtypes: datetime64[us](1), float64(2), int64(1), object(3), str(1)
memory usage: 32.1+ MB


### Interpretation

- The dataset contains 525,461 rows and 8 columns.
- The `Customer ID` column has approximately 20.5% missing values.
- The `Description` column has less than 1% missing values.
- `InvoiceDate` is already stored as a datetime type, so no conversion is needed.

## 6. Summary Statistics

Generate summary statistics for numerical columns to understand their distribution and identify potential anomalies.

In [21]:
df.describe()

,Quantity,InvoiceDate,Price,Customer ID
count,525461.000000,525461,525461.000000,417534.000000
mean,10.337667,2010-06-28 11:37:36.845018,4.688834,15360.645478
min,-9600.000000,2009-12-01 07:45:00,-53594.360000,12346.000000
25%,1.000000,2010-03-21 12:20:00,1.250000,13983.000000
50%,3.000000,2010-07-06 09:51:00,2.100000,15311.000000
75%,10.000000,2010-10-15 12:45:00,4.210000,16799.000000
max,19152.000000,2010-12-09 20:01:00,25111.090000,18287.000000
std,107.424110,NaN,146.126914,1680.811316


### Interpretation

- `Quantity` and `Price` contain a wide range of values, suggesting the presence of potential outliers.
- `Price` includes negative values, which require further investigation during the data cleaning phase.
- `Customer ID` is treated as a numerical column by Pandas, but it represents an identifier rather than a measurable quantity.

## 7. Missing Values Overview

Assess the amount of missing data in each column before deciding how to handle it.

In [22]:
missing_count = df.isnull().sum()
missing_percentage = (df.isnull().sum() / len(df)) * 100

missing_values = pd.DataFrame({
    "Missing Count": missing_count,
    "Missing Percentage": missing_percentage
})

missing_values

,Missing Count,Missing Percentage
Invoice,0,0.000000
StockCode,0,0.000000
Description,2928,0.557225
Quantity,0,0.000000
InvoiceDate,0,0.000000
Price,0,0.000000
Customer ID,107927,20.539488
Country,0,0.000000


### Interpretation

- The `Customer ID` column contains a considerable number of missing values. Its impact will depend on the type of analysis performed later.
- The `Description` column has a small number of missing values, but this column is essential for product-level analysis and will require careful handling during the data cleaning stage.
- No decisions about removing or imputing missing values are made in this notebook.

## 8. Duplicate Rows

Check whether the dataset contains duplicated records before performing any data cleaning.

In [23]:
df.duplicated().sum()

duplicates = df[df.duplicated(keep=False)]
duplicates.sort_values(by=["Invoice", "StockCode"]).head(20)

,Invoice,StockCode,Description,Quantity,InvoiceDate,Price,Customer ID,Country
379,489517,21491,SET OF THREE VINTAGE GIFT WRAPS,1,2009-12-01 11:34:00,1.95,16329.0,United Kingdom
391,489517,21491,SET OF THREE VINTAGE GIFT WRAPS,1,2009-12-01 11:34:00,1.95,16329.0,United Kingdom
365,489517,21821,GLITTER STAR GARLAND WITH BELLS,1,2009-12-01 11:34:00,3.75,16329.0,United Kingdom
386,489517,21821,GLITTER STAR GARLAND WITH BELLS,1,2009-12-01 11:34:00,3.75,16329.0,United Kingdom
363,489517,21912,VINTAGE SNAKES & LADDERS,1,2009-12-01 11:34:00,3.75,16329.0,United Kingdom
371,489517,21912,VINTAGE SNAKES & LADDERS,1,2009-12-01 11:34:00,3.75,16329.0,United Kingdom
394,489517,21912,VINTAGE SNAKES & LADDERS,1,2009-12-01 11:34:00,3.75,16329.0,United Kingdom
362,489517,21913,VINTAGE SEASIDE JIGSAW PUZZLES,1,2009-12-01 11:34:00,3.75,16329.0,United Kingdom
385,489517,21913,VINTAGE SEASIDE JIGSAW PUZZLES,1,2009-12-01 11:34:00,3.75,16329.0,United Kingdom
368,489517,22130,PARTY CONE CHRISTMAS DECORATION,6,2009-12-01 11:34:00,0.85,16329.0,United Kingdom


### Interpretation

- The dataset contains **6,865 fully duplicated records**.
- Inspection of duplicated groups confirms that these records are identical across all columns.
- No duplicate records are removed at this stage. The decision will be implemented and documented in the data cleaning notebook.

## 9. Unique Values Overview

Explore the categorical columns to understand their diversity and identify potential data quality issues.

In [24]:
df.nunique()

Invoice        28816
StockCode       4632
Description     4681
Quantity         825
InvoiceDate    25296
Price           1606
Customer ID     4383
Country           40
dtype: int64

### Interpretation

- The dataset contains transactions from **40 countries**, allowing future geographical analysis.
- The number of unique invoices is much smaller than the number of rows, indicating that each invoice may contain multiple products.
- The number of unique `Description` values is greater than the number of unique `StockCode` values. This suggests that at least some products may have multiple descriptions, requiring further investigation during the exploratory analysis stage.

## Initial Data Quality Assessment

The initial inspection revealed several data quality issues that should be addressed before analysis:

- Missing values in `Customer ID` and `Description`
- Fully duplicated records
- Negative values in the `Price` column
- `Customer ID` stored as float instead of an identifier
- Differences between the number of unique `StockCode` and `Description` values that require further investigation